extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/config/GrpcConfigBuilder.java

```diff
@@ -0,0 +1,61 @@
+package io.quarkus.grpc.runtime.config;
+
+import static io.vertx.core.net.SocketAddress.inetSocketAddress;
+import static java.lang.Integer.MAX_VALUE;
+import static java.net.InetAddress.getByName;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ConfigBuilder;
+import io.smallrye.config.ConfigSourceContext;
+import io.smallrye.config.ConfigSourceFactory.ConfigurableConfigSourceFactory;
+import io.smallrye.config.SmallRyeConfigBuilder;
+import io.smallrye.config.common.MapBackedConfigSource;
+import io.vertx.core.net.SocketAddress;
+
+public class GrpcConfigBuilder implements ConfigBuilder {
+    @Override
+    public SmallRyeConfigBuilder configBuilder(final SmallRyeConfigBuilder builder) {
+        return builder.withSources(new RandomPortConfigSourceFactory());
+    }
+
+    private static class RandomPortConfigSourceFactory implements ConfigurableConfigSourceFactory<GrpcServerConfig> {
+        @Override
+        public Iterable<ConfigSource> getConfigSources(final ConfigSourceContext context, final GrpcServerConfig config) {
+            Map<String, String> randomPorts = new HashMap<>();
+
+            String host = config.host();
+            if (LaunchMode.current().equals(LaunchMode.TEST)) {
+                if (config.testPort() == 0) {
```

Comment:
No that is not correct,

Port 0 is a random port, but everything with port 0 uses the same
Port -1 is a random port, but everything with port -1 uses the same
...

Reply:
and that's across all extensions.

Reply:
I didn't notice that the client was in a separate group. Let me fix that.